### PR TITLE
[Perf] Use SDPA for BLIP-2 Q-Former attention

### PR DESCRIPTION
**Pull request by Levius-Fubuki**
## Purpose

BLIP-2 Q-Former attention currently materializes the attention scores, applies
scaling and softmax separately, and launches a second matrix multiplication:

```text
before: {matmul: 2, mul: 1, softmax: 1, dropout: 1}
after:  {scaled_dot_product_attention: 1}
```

This PR uses standard `torch.nn.functional.scaled_dot_product_attention` for
both Q-Former self-attention and cross-attention. It preserves the existing
scale and training dropout behavior while allowing PyTorch to dispatch fused
attention kernels.

This removes the score-sized attention intermediate and the separate scale,
softmax, dropout, and value-matmul launches. It keeps one portable PyTorch path
with no custom kernel, architecture gate, device branch, or device-specific
tuning.

I searched the open PRs for BLIP-2, Q-Former, and SDPA changes and found no
active PR addressing this attention path.

### Lint

```bash
ruff check vllm/model_executor/models/blip2.py
ruff format --check vllm/model_executor/models/blip2.py
git diff --check origin/main..HEAD
```

### RTX 4090 default eager encoder benchmark

The performance harness constructs the real
`Blip2ForConditionalGeneration` class with BLIP-2-compatible dimensions and
invokes its real `embed_multimodal` boundary. Both variants use identical,
finite, deterministic weights.

Hardware/software: NVIDIA RTX 4090, CUDA 13.0, PyTorch 2.13.0+cu130.
Timing uses 25 warmups, 200 iterations, and five alternating repeats.

| Dtype | Batch | Scope | Old p50 (us) | New p50 (us) | Speedup |
|---|---:|---|---:|---:|---:|
| FP16 | 1 | Q-Former | 1519.28 | 1049.77 | 1.45x |
| FP16 | 4 | Q-Former | 1521.30 | 1006.97 | 1.51x |
| BF16 | 1 | Q-Former | 1504.88 | 1052.73 | 1.43x |
| BF16 | 4 | Q-Former | 1565.29 | 1088.26 | 1.44x |
| FP16 | 1 | `embed_multimodal` | 2797.99 | 2322.34 | 1.20x |
| FP16 | 4 | `embed_multimodal` | 2744.05 | 2196.03 | 1.25x |
| BF16 | 1 | `embed_multimodal` | 2691.73 | 2248.67 | 1.20x |
| BF16 | 4 | `embed_multimodal` | 2748.79 | 2177.32 | 1.26x |

The Q-Former geometric-mean speedup is 1.48x for FP16 and 1.43x for BF16.
All eight rows passed numerical, finite-value, repeatability, shape, and
input-mutation validation with no performance regression.

Q-Former peak extra allocation decreased from 1,579,008 to 1,237,504 bytes at
batch 1, and from approximately 7.3 MiB to 5.1 MiB at batch 4.

Profiler results replace 80 `aten::bmm` calls and 40 separate softmax calls per
ten Q-Former invocations with 40 `aten::scaled_dot_product_attention` calls.
All candidate calls dispatch to `aten::_flash_attention_forward`.

The harness, raw CSV, gate result, and profiler summary are available in the
[benchmark artifact Gist](https://gist.github.com/Levius-Fubuki/18c576619c30d983466d2f57cd6ee7e6).

## Test Result

`ruff-check`, `ruff-format`, and `git diff --check` passed.

## Limitations

These results come from one RTX 4090 run with deterministic weights, not a
full pretrained BLIP-2 checkpoint or end-to-end request throughput benchmark.

Current BLIP-2 does not implement `SupportsEncoderCudaGraph`, and runtime hooks
confirmed that Q-Former executes on the default eager multimodal encoder path.
No CUDA Graph performance claim is made. If BLIP-2 gains encoder CUDA Graph
support later, batch-1 short-key shapes should be benchmarked again before
enabling capture.

No performance claim is made for other architectures, devices, or PyTorch
versions. Full upstream cross-backend CI still requires a maintainer `ready`
or `verified` label.

## AI assistance

AI assistance was used during implementation, testing, benchmarking, and PR
drafting. The human contributor reviewed and takes responsibility for the
change and evidence.

### CODE REVIEW

File: vllm/model_executor/models/blip2.py [hunk 1]
```diff
@@ -141,14 +142,13 @@ def forward(
...
 
-        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
-        attention_probs = torch.softmax(attention_scores * self.scaling, dim=-1)
-
-        # This is actually dropping out entire tokens to attend to, which might
-        # seem a bit unusual, but is taken from the original Transformer paper.
-        attention_probs_dropped = self.dropout(attention_probs)
-
-        context_layer = torch.matmul(attention_probs_dropped, value_layer)
+        context_layer = F.scaled_dot_product_attention(
```
Comment (DarkLight1337) [inline vllm/model_executor/models/blip2.py:145]: Is this in the Transformers code? cc @hmellor

Comment (hmellor) [reply vllm/model_executor/models/blip2.py:145]: Not really, but it could.

`Blip2QFormerMultiHeadAttention` in Transformers uses the `ALL_ATTENTION_FUNCTIONS` registry but the `Blip2QFormerConfig` does not specify `_attn_implementation`. So, it falls back to `eager_attention_forward`:

```py
def eager_attention_forward(
    module: nn.Module,
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attention_mask: torch.Tensor | None,
    scaling: float,
    dropout: float = 0.0,
    **kwargs,
):
    attn_weights = torch.matmul(query, key.transpose(-1, -2)) * scaling
    if attention_mask is not None:
        attn_weights = attn_weights + attention_mask

    attn_weights = nn.functional.softmax(attn_weights, dim=-1)
    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)

    attn_output = torch.matmul(attn_weights, value)
    attn_output = attn_output.transpose(1, 2).contiguous()

    return attn_output, attn_weights
```

However if the config had been configured with `"sdpa"` then it would have called `sdpa_attention_forward` in Transformers.

Comment (DarkLight1337) [reply vllm/model_executor/models/blip2.py:145]: I wonder whether we could simply use vllm-native `MMEncoderAttention` for this? @Isotr0py

Comment (Isotr0py) [reply vllm/model_executor/models/blip2.py:145]: I think we can use `MMEncoderAttention` since it's just a normal bidirectional attention here.

Comment (Levius-Fubuki) [reply vllm/model_executor/models/blip2.py:145]: Would `MMEncoderAttention` also support the Q-Former cross-attention case where `q_len != kv_len` (e.g. 32 vs 257)? I may be missing something, but the current Flash/Triton wrappers seem to reuse sequence metadata derived from `q_len` for K/V.

Comment (Isotr0py) [reply vllm/model_executor/models/blip2.py:145]: I see, let's use `F.sdpa` here then

File: tests/models/multimodal/test_blip2_qformer_attention.py
Comment (Isotr0py) [inline tests/models/multimodal/test_blip2_qformer_attention.py:1]: Please remove this redundant test

**Review by Isotr0py — APPROVED** [review]

**Comment by Levius-Fubuki** [comment]
Hi @Isotr0py, the full Buildkite CI has passed. It looks like the earlier `pre-run-check` failure is still blocking the PR because it ran before the `ready` label was added. Would you mind rerunning it when convenient?

**Comment by DarkLight1337** [comment]
Retrying
